db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java

```diff
@@ -26,348 +26,400 @@
-		}
-		if (!ExecutorUtils.shutdownNowAndAwaitTermination(updateHeartbeatExecutor, Duration.ofSeconds(5))) {
-			LOG.warn("Failed to shutdown update-heartbeat-executor properly.");
+  private static final Duration DEFAULT_SHUTDOWN_TIMEOUT = Duration.ofMinutes(30);
+  private static final Duration THREAD_POOL_TIMEOUT = Duration.ofMinutes(5);
+
+  public static final double TRIGGER_NEXT_BATCH_WHEN_AVAILABLE_THREADS_RATIO = 0.5;
+  public static final String THREAD_PREFIX = "db-scheduler";
+  private static final Logger LOG = LoggerFactory.getLogger(Scheduler.class);
+  private final SchedulerClient delegate;
+  private final Clock clock;
+  private final TaskRepository taskRepository;
+  private final TaskResolver taskResolver;
+  private final int parallelism;
+  private final Waiter executeDueWaiter;
+  private final Duration deleteUnresolvedAfter;
+  protected final List<OnStartup> onStartup;
+  private final Waiter detectDeadWaiter;
+  private final Duration heartbeatInterval;
+  private final StatsRegistry statsRegistry;
+  private final int pollingLimit;
+
+  private final ExecutorService dueExecutor;
+  private final ExecutorService detectDeadExecutor;
+  private final ExecutorService updateHeartbeatExecutor;
+
+  private final ExecutorService executorService;
+	private final boolean managedExecutorService; //if true, the executorService is managed by this instance
+  public final Duration executorShutdownWait;
+
+  private final Map<Execution, CurrentlyExecuting> currentlyProcessing = Collections.synchronizedMap(new HashMap<>());
+  private final Waiter heartbeatWaiter;
+  private final SettableSchedulerState schedulerState = new SettableSchedulerState();
+  private int currentGenerationNumber = 1;
+
+  protected Scheduler(Clock clock, TaskRepository taskRepository, TaskResolver taskResolver, int parallelism, ExecutorService executorService, Duration executorShutdownWait, SchedulerName schedulerName,
+                      Waiter executeDueWaiter, Duration heartbeatInterval, boolean enableImmediateExecution, StatsRegistry statsRegistry, int pollingLimit, Duration deleteUnresolvedAfter, List<OnStartup> onStartup) {
+    this.clock = clock;
+    this.taskRepository = taskRepository;
+    this.taskResolver = taskResolver;
+    this.parallelism = parallelism;
+    this.executeDueWaiter = executeDueWaiter;
+    this.deleteUnresolvedAfter = deleteUnresolvedAfter;
+    this.onStartup = onStartup;
+    this.detectDeadWaiter = new Waiter(heartbeatInterval.multipliedBy(2), clock);
+    this.heartbeatInterval = heartbeatInterval;
+    this.heartbeatWaiter = new Waiter(heartbeatInterval, clock);
+    this.statsRegistry = statsRegistry;
+    this.pollingLimit = pollingLimit;
+
+    ThreadPoolExecutor housekeepingExecutor = newThreadPool(THREAD_PREFIX + "-housekeeper-", 3, THREAD_POOL_TIMEOUT);
+    this.dueExecutor = housekeepingExecutor;
+    this.detectDeadExecutor = housekeepingExecutor;
+    this.updateHeartbeatExecutor = housekeepingExecutor;
```

Comment:
Not sure about this. The util-threads currently run "forever" (internal sleep). The only difference now will be that the thread-names will be a bit less describing. What do you feel is the benefit here?

Maybe they should be made into a `ScheduledExecutorService`, but I suspect that generates a bigger change..